crates/iroha_primitives/src/json.rs

```diff
@@ -131,28 +104,239 @@ impl FromStr for JsonString {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         if let Ok(value) = serde_json::from_str::<Value>(s) {
-            Ok(JsonString(value.to_string()))
+            Ok(JsonString(value))
         } else {
             let json_formatted_string = serde_json::to_string(s)?;
             let value: Value = serde_json::from_str(&json_formatted_string)?;
-            Ok(JsonString(value.to_string()))
+            Ok(JsonString(value))
         }
     }
 }
 
 impl Default for JsonString {
     fn default() -> Self {
-        // NOTE: empty string isn't valid JSON
-        Self("null".to_string())
+        Self(Value::Null)
+    }
+}
+
+impl Ord for JsonString {
+    fn cmp(&self, other: &Self) -> Ordering {
+        JsonStringRef(&self.0).cmp(&JsonStringRef(&other.0))
     }
 }
 
-impl AsRef<str> for JsonString {
-    fn as_ref(&self) -> &str {
-        &self.0
+impl PartialOrd for JsonString {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
     }
 }
 
+impl PartialEq for JsonString {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other).is_eq()
+    }
+}
+
+impl Eq for JsonString {}
+
+impl Ord for JsonStringRef<'_> {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        self.0.to_string().cmp(&other.0.to_string())
```

Comment:
I'm not a fan of converting to string every time

Reply:
you could match on some of the variants of `serde_json::Value` and compare

Reply:
We can compare everything except numbers.
f64 doesn't implement Ord, but we can compare numbers as strings